### PR TITLE
HLS support

### PR DESCRIPTION
**Pull request by meyskens**
- [x] Add a pool for HLS frames
- [x] Make the pool clean itself
- [x] ~~Add metadata~~
- [x] ~~Look into extended metadata~~
- [x] Add listener statistics (?listener=id on segments with a ttl to track)

Suggestions/tips are always welcome

### CODE REVIEW

File: intern/streams/httphandler.js [hunk 1]
```diff
@@ -62,6 +62,30 @@ var httpHandler = function(app) {
...
+        if (!global.streams.streamExists(req.params.stream)) {
+            res.status(404).send("Stream not found")
+            return
+        }
+        var streamConf = global.streams.getStreamConf(req.params.stream)
+
+        // generate response header
+        var headers = {
+            "Content-Type": streamConf.type || "audio/mpeg",
+            "Connection": 'close',
```
Comment (leoetlino) [inline intern/streams/httphandler.js]: should use double quotes

File: intern/streams/httphandler.js [hunk 2]
```diff
@@ -62,6 +62,30 @@ var httpHandler = function(app) {
...
+            "Access-Control-Allow-Origin": "*",
+            "X-Begin": "Thu, 30 Jan 2014 17:20:00 GMT",
+            "Cache-Control": "no-cache",
+            "Expires": "Sun, 9 Feb 2014 15:32:00 GMT"
+        }
+        if (!global.streams.hlsPool[req.params.stream][req.params.segment]){
+                return res.status(404).send("Segment not found")
```
Comment (leoetlino) [inline intern/streams/httphandler.js]: indentation

File: intern/streams/httphandler.js (hunk 1 above)
Comment (meyskens) [inline intern/streams/httphandler.js]: Agreed, the formatation rules are not yet in place on Cast. They will be enforced in all parts with the refactor. I'll change them in the new parts when i work on them.

File: intern/streams/httphandler.js [hunk 3]
```diff
@@ -216,6 +244,21 @@ var httpHandler = function(app) {
         res.send(global.config.hostname + "/streams/" + stream)
     }
 
+    var serveM3U8 = function(req, res) {
+        var stream = req.params[0].split(".")[0]
+        if (typeof stream === "undefined" || stream === "" || !global.streams.streamExists(stream)) {
```
Comment (leoetlino) [inline intern/streams/httphandler.js]: `!stream` is simpler and also tests for `null` (even if that is unlikely to be a possible value)

File: intern/streams/httphandler.js (hunk 2 above)
Comment (meyskens) [inline intern/streams/httphandler.js]: :+1:

File: intern/streams/httphandler.js [hunk 4]
```diff
@@ -216,6 +244,21 @@ var httpHandler = function(app) {
...
+            res.status(404).send("Stream not found")
+            return
+        }
+
+        res.setHeader("Content-Type", "audio/x-mpegurl")
+        var response = "#EXTM3U\n#EXT-X-VERSION:5\n#EXT-X-TARGETDURATION:5\n"
+        for (var id in global.streams.hlsIndexes[stream]){
```
Comment (leoetlino) [inline intern/streams/httphandler.js]: space before brace

File: intern/streams/httphandler.js (hunk 3 above)
Comment (meyskens) [inline intern/streams/httphandler.js]: Yet another part that will be fixed in the refactor.

File: intern/streams/httphandler.js [hunk 5]
```diff
@@ -216,6 +244,21 @@ var httpHandler = function(app) {
...
+            response+="#EXTINF:5.0,\n" + global.config.hostname + "/hls/"+stream+"/"+global.streams.hlsIndexes[stream][id]+"\n"
```
Comment (leoetlino) [inline intern/streams/httphandler.js]: space before and after `+=` and `+`

File: intern/streams/streams.js [hunk 6]
```diff
@@ -68,6 +72,34 @@ var addStream = function(inputStream, conf) {
         inputStream.pipe(throttleStream);
     }
 
+    if (global.config.hls){
```
Comment (leoetlino) [inline intern/streams/streams.js]: space before brace

**Comment by meyskens** [comment]
Metadata to be done by https://github.com/Innovate-Technologies/id3-writer-js

**Comment by meyskens** [comment]
Tracking added by adding ?id=x to the URL. The TTL for HLS listeners is 20 seconds (= 4 segments).

**Comment by meyskens** [comment]
Metadata is not present in this release, the ID3 tags give playback issues.

**Comment by meyskens** [comment]
Adaptive streaming is under consideration, this might give issues with some encoders having delay. We'll do some tests soon to see what this does.
